fuel/utils.py

```diff
@@ -14,6 +16,390 @@
+            raise ValueError("trying to add two Subset instances with "
+                             "different numbers of original examples, they "
+                             "can't possibly belong to the same dataset")
+        # An empty subset is a neutral element in subset algebra
+        if self.is_empty:
+            return other
+        # Merging list-based and slice-based subsets results in a list
+        # conversion
+        if self.is_list != other.is_list:
+            return self.__class__(
+                self.get_list_representation() +
+                other.get_list_representation(),
+                self.original_num_examples)
+        # List-based subsets are merged by concatenating their indices.
+        if self.is_list:
+            return self.__class__(self.list_or_slice + other.list_or_slice,
+                                  self.original_num_examples)
+        # Slice-based subsets are merged into a slice-based subset if they
+        # overlap, otherwise they're converted to a list-based subset.
+        else:
+            self_sss = self.slice_to_numerical_args(
+                self.list_or_slice, self.original_num_examples)
+            self_start, self_stop, self_step = self_sss
+            other_sss = self.slice_to_numerical_args(
+                other.list_or_slice, other.original_num_examples)
+            other_start, other_stop, other_step = other_sss
+            # In case of overlap, the solution is to choose the smallest start
+            # value and largest stop value.
+            if not (self_stop < other_start or self_start > other_stop):
+                return self.__class__(slice(min(self_start, other_start),
+                                            max(self_stop, other_stop),
+                                            self_step),
+                                      self.original_num_examples)
+            # Everything else is transformed into lists before merging.
+            else:
+                return self.__class__(
+                    self.get_list_representation() +
+                    other.get_list_representation(),
+                    self.original_num_examples)
+
+    def __getitem__(self, key):
+        """Translates a request from this subset to the dataset.
+
+        A request made in the context of this subset is translated into a
+        request on the dataset itself.
+
+        Parameters
+        ----------
+        key : :class:`list` or :class:`slice`
+            A request made *within the context of this subset*.
+
+        Returns
+        -------
+        :class:`list` or :class:`slice`
+            The translated request to be used on the dataset.
+
+        """
+        self._request_sanity_check(key, self.num_examples)
+        # slice(None, None, None) selects the whole subset, no need to index
+        # anything
+        if key == slice(None, None, None):
+            return self.list_or_slice
+        elif self._is_list(key):
```

Comment:
Maybe it is a matter of state, but I think that all this extra indentation makes understanding harder compared to a list of `if condition: return smth` statements.